libcontainer/notify_linux_v2.go

```diff
@@ -0,0 +1,127 @@
+// +build linux
+
+package libcontainer
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"unsafe"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+)
+
+func getOOMCount(path string) (int, error) {
+	if _, err := os.Lstat(path); err != nil {
+		return 0, err
+	}
+
+	content, err := ioutil.ReadFile(path)
+	if err != nil {
+		return 0, err
+	}
+
+	oomCount := 0
+	lines := strings.Split(string(content), "\n")
+	for _, line := range lines {
+		arr := strings.Split(line, " ")
+		if len(arr) == 2 && (arr[0] == "oom" || arr[0] == "oom_kill") {
+			count, err := strconv.Atoi(arr[1])
+			if err == nil && count > oomCount {
+				oomCount = count
+			}
+		}
+	}
+	return oomCount, nil
+}
+
+func getPopulated(path string) (int, error) {
+	if _, err := os.Lstat(path); err != nil {
+		return 0, err
+	}
+
+	content, err := ioutil.ReadFile(path)
+	if err != nil {
+		return 0, err
+	}
+	lines := strings.Split(string(content), "\n")
+	for _, line := range lines {
+		arr := strings.Split(line, " ")
+		if len(arr) == 2 && arr[0] == "populated" {
+			return strconv.Atoi(arr[1])
+		}
+	}
+	return 0, nil
+}
+
+func registerMemoryEventV2(cgDir string, evName string, cgEvName string) (<-chan struct{}, error) {
+	eventControlPath := filepath.Join(cgDir, evName)
+	cgEvPath := filepath.Join(cgDir, cgEvName)
+	fd, err := unix.InotifyInit()
+	if err != nil {
+		return nil, err
+	}
+	// watching oom kill
+	_, err = unix.InotifyAddWatch(fd, eventControlPath, unix.IN_MODIFY)
+	if err != nil {
+		unix.Close(fd)
+		return nil, err
+	}
+	// Because no `unix.IN_DELETE|unix.IN_DELETE_SELF` event for cgroup file system, so watching all process exited
+	_, err = unix.InotifyAddWatch(fd, cgEvPath, unix.IN_MODIFY)
+	if err != nil {
+		unix.Close(fd)
+		return nil, err
+	}
+	ch := make(chan struct{})
+	go func() {
+		var (
+			buffer [unix.SizeofInotifyEvent + unix.PathMax]byte
+			offset uint32
+		)
+		defer func() {
+			unix.Close(fd)
+			close(ch)
+		}()
+
+		for {
+			n, err := unix.Read(fd, buffer[:])
+			if err != nil {
+				logrus.Warn(err)
+				return
+			}
+			offset = 0
+			for offset <= uint32(n-unix.SizeofInotifyEvent) {
+				rawEvent := (*unix.InotifyEvent)(unsafe.Pointer(&buffer[offset]))
+				bytes := (*[unix.PathMax]byte)(unsafe.Pointer(&buffer[unix.SizeofInotifyEvent]))
+				eventName := strings.TrimRight(string(bytes[0:rawEvent.Len]), "\000")
```

Comment:
1. This is actually a file name, not an event name.

2. It might be easier to just save the two fds returned by `InotifyAddWatch` above and check if `rawEvent.Wd` is one of them. Let me take a look...

Reply:
That's right. You can just check the fd, no need to check the file name. Simpler code and faster checks.

```go
evFd, err := unix.InotifyAddWatch(fd, eventControlPath, unix.IN_MODIFY)
...
cgFd, err := unix.InotifyAddWatch(fd, cgEvPath, unix.IN_MODIFY)
...
for offset <= uint32(n-unix.SizeofInotifyEvent) {
  event := (*unix.InotifyEvent)(unsafe.Pointer(&buffer[offset]))
  offset += unix.SizeofInotifyEvent + uint32(event.Len)
  if event.Mask&unix.IN_MODIFY != unix.IN_MODIFY {
    continue
  }
  switch int(event.Wd) {
    case evFd:
      oom, err := getOOMCount(eventControlPath)
      ...
    case cgFd:
      pids, err := getPopulated(cgEvPath)
...
```